### DISCUSSION TITLE
RAG API failed to start with railway

### DISCUSSION THREAD

**Discussion by bloodchen** [danny-avila/LibreChat#3530] [Troubleshooting] [answered]
### What happened?

Just use the default deploy, default variables . The RAG API can not start. 

### Steps to Reproduce

1. deploy to railway 2. deploy 3. see the errors

### Relevant log output

```shell
Starting Container

2024-08-04 09:46:59,940 - root - INFO - Initialized embeddings of type: <class 'langchain_openai.embeddings.base.OpenAIEmbeddings'>

/usr/local/lib/python3.10/site-packages/langchain_community/vectorstores/pgvector.py:322: LangChainPendingDeprecationWarning: Please use JSONB instead of JSON for metadata. This change will allow for more efficient querying that involves filtering based on metadata.Please note that filtering operators have been changed when using JSOB metadata to be prefixed with a $ sign to avoid name collisions with columns. If you're using an existing database, you will need to create adb migration for your metadata column to be JSONB and update your queries to use the new operators. 

  warn_deprecated(

2024-08-04 09:47:00,196 - uvicorn.error - INFO - Started server process [1]

2024-08-04 09:47:00,196 - uvicorn.error - INFO - Waiting for application startup.

2024-08-04 09:47:00,196 - uvicorn.error - INFO - Application startup complete.

2024-08-04 09:47:00,197 - uvicorn.error - INFO - Uvicorn running on http://0.0.0.0:8000 (Press CTRL+C to quit)
```

### Screenshots

![image](https://github.com/user-attachments/assets/21700015-d626-4d19-ac26-037de2712c55)

### Code of Conduct

- [X] I agree to follow this project's Code of Conduct

**Comment by fredfortin1** [comment]
Exactly the same error on my end

**Comment by ctwhome** [comment]
same issue here: #3511

**Comment by vcavallo** [comment]
Same for me

**Comment by monotykamary** [comment, marked as answer]
For anyone who is having trouble with Railway, the workaround is to set a `PORT` for the RAG API service to 8000.
```
PORT=8000
```

Not sure when, but in Railway's [documentation](https://docs.railway.app/guides/public-networking#railway-provided-domain), it mentions this:

> If you prefer to explicitly set a port, you must set the PORT variable in your service variables to the port on which your service is listening.

And I've also updated the `RAG_API_URL` in the `LibreChat` service to `https://${{" RAG API 📚".RAILWAY_PUBLIC_DOMAIN}}`, as it seems to return an empty `http://` first time around.
```
RAG_API_URL=https://${{" RAG API 📚".RAILWAY_PUBLIC_DOMAIN}}
```

cc @danny-avila 

EDIT:
Here is my updated template: https://railway.app/template/HxvQtm?referralCode=ZqgrJ0

**Reply by larryb821** [reply]
Thank you! the port=8000 variable solved the issue for me on RAG deployment!

**Reply by FernandoLLB** [reply]
Just to be clear I have to change the DB_PORT to 8000 right?
Or do I have to create a new PORT=8000 variable?

**Reply by akilicarslan** [reply]
> https://railway.app/template/HxvQtm?referralCode=ZqgrJ0

the new template is not working for me. PORT=8000 is not working as well.

Edit: I think it is about Railway dynamic variables. I used static values instead of Railway dynamic variables. It worked after that. I changed RAG_API_URL to static url.

**Reply by monotykamary** [reply]
I've also left an issue on Railway on their dynamic variables, but it couldn't be reproduced when their team checked. It's like a heisenbug sometimes.

**Reply by berry-13** [reply]
btw, the [official LibreChat template](https://www.librechat.ai/docs/remote/railway) has supported RAG for quite a while. Without using it, we won't be able to assist you in troubleshooting any future issues you might encounter

**Comment by FernandoLLB** [comment]
Just to be clear I have to change the DB_PORT to 8000 right?
Or do I have to create a new PORT=8000 variable?

**Reply by berry-13** [reply]
> Just to be clear I have to change the DB_PORT to 8000 right? Or do I have to create a new PORT=8000 variable?

no, you need to set back the DB_PORT to the default one and add a new one with PORT=8000. You're already using the template right? since I updated it 1 month ago

**Reply by FernandoLLB** [reply]
Yes I am using the template and it solved the RAG issue but know when I set API KEYS for the endpoints it gives me errors also this time related to MEILISEARCH.
